### ISSUE TITLE
INP可视化

### ISSUE THREAD

**Issue by Harry-Bai** [luow23/INP-Former#28]
作者您好，想请教一下如何可视化提取到的INP呀

**Comment by luow23** [comment]
嗨咯，您好，INP可视化是利用注意力里面的交叉注意力图得到的，我们是提取的第四层的交叉注意力图作为可视化

**Comment by Harry-Bai** [comment]
> 嗨咯，您好，INP可视化是利用注意力里面的交叉注意力图得到的，我们是提取的第四层的交叉注意力图作为可视化

谢谢~

**Comment by Harry-Bai** [comment]
您如果方便的话可以提供下INP的可视化代码作为参考吗

**Comment by luow23** [comment]
嗨咯，您好，这是我们保存注意力图的代码
def save_imag(imgs, anomaly_map, gt, prototype_map, attention_map, save_root, img_path):
    batch_num = imgs.shape[0]
    for i in range(batch_num):
        img_path_list = img_path[i].split('\\')
        class_name, category, idx_name = img_path_list[-4], img_path_list[-2], img_path_list[-1]
        # print(os.path.join(save_root, class_name, category))
        os.makedirs(os.path.join(save_root, class_name, category), exist_ok=True)
        input_frame = denormalize(imgs[i].clone().squeeze(0).cpu().detach().numpy())
        cv2_input = np.array(input_frame, dtype=np.uint8)
        plt.imsave(os.path.join(save_root, class_name, category, fr'{idx_name}_0.png'), cv2_input)
        ano_map = anomaly_map[i].squeeze(0).cpu().detach().numpy()
        plt.imsave(os.path.join(save_root, class_name, category, fr'{idx_name}_1.png'), ano_map, cmap='jet')
        gt_map = gt[i].squeeze(0).cpu().detach().numpy()
        plt.imsave(os.path.join(save_root, class_name, category, fr'{idx_name}_2.png'), gt_map, cmap='gray')
        distance = prototype_map[i].view((28, 28)).cpu().detach().numpy()
        distance = cv2.resize(distance, (392, 392), interpolation=cv2.INTER_AREA)
        plt.imsave(os.path.join(save_root, class_name, category, fr'{idx_name}_3.png'), distance, cmap='jet')
        attention = attention_map[i]

        prototype_num = attention.shape[0]
        for j in range(prototype_num):
            atten = attention[j]
            max_value = max(atten)
            min_value = min(atten)
            atten = (atten-min_value)/(max_value-min_value)
            atten = atten.view((28, 28)).cpu().detach().numpy()
            index = 3 + j + 1
            atten = cv2.resize(atten, (392, 392), interpolation=cv2.INTER_AREA)
            plt.imsave(os.path.join(save_root, class_name, category, fr'{idx_name}_{index}.png'), atten, cmap='jet')
        attention_sum = torch.sum(attention, dim=0).view((28, 28)).cpu().detach().numpy()
        plt.imsave(os.path.join(save_root, class_name, category, fr'{idx_name}_{4+prototype_num}.png'), attention_sum, cmap='hot')
        plt.close()

**Comment by Harry-Bai** [comment]
> 嗨咯，您好，这是我们保存注意力图的代码
> 
> ```
>    batch_num = imgs.shape[0]
>    for i in range(batch_num):
>        img_path_list = img_path[i].split('\\')
>        class_name, category, idx_name = img_path_list[-4], img_path_list[-2], img_path_list[-1]
>        # print(os.path.join(save_root, class_name, category))
>        os.makedirs(os.path.join(save_root, class_name, category), exist_ok=True)
>        input_frame = denormalize(imgs[i].clone().squeeze(0).cpu().detach().numpy())
>        cv2_input = np.array(input_frame, dtype=np.uint8)
>        plt.imsave(os.path.join(save_root, class_name, category, fr'{idx_name}_0.png'), cv2_input)
>        ano_map = anomaly_map[i].squeeze(0).cpu().detach().numpy()
>        plt.imsave(os.path.join(save_root, class_name, category, fr'{idx_name}_1.png'), ano_map, cmap='jet')
>        gt_map = gt[i].squeeze(0).cpu().detach().numpy()
>        plt.imsave(os.path.join(save_root, class_name, category, fr'{idx_name}_2.png'), gt_map, cmap='gray')
>        distance = prototype_map[i].view((28, 28)).cpu().detach().numpy()
>        distance = cv2.resize(distance, (392, 392), interpolation=cv2.INTER_AREA)
>        plt.imsave(os.path.join(save_root, class_name, category, fr'{idx_name}_3.png'), distance, cmap='jet')
>        attention = attention_map[i]
>        prototype_num = attention.shape[0]
>        for j in range(prototype_num):
>            atten = attention[j]
>            max_value = max(atten)
>            min_value = min(atten)
>            atten = (atten-min_value)/(max_value-min_value)
>            atten = atten.view((28, 28)).cpu().detach().numpy()
>            index = 3 + j + 1
>            atten = cv2.resize(atten, (392, 392), interpolation=cv2.INTER_AREA)
>            plt.imsave(os.path.join(save_root, class_name, category, fr'{idx_name}_{index}.png'), atten, cmap='jet')
>        attention_sum = torch.sum(attention, dim=0).view((28, 28)).cpu().detach().numpy()
>        plt.imsave(os.path.join(save_root, class_name, category, fr'{idx_name}_{4+prototype_num}.png'), attention_sum, cmap='hot')
>        plt.close()
> ```

非常感谢^_^

**Comment by luow23** [comment]
你好，我们用的解码器注意力图来可视化INP

**Comment by jieduring** [comment]
好的，感谢!



---原始邮件---
发件人: "Wei ***@***.***&gt;
发送时间: 2025年5月23日(周五) 上午10:44
收件人: ***@***.***&gt;;
抄送: ***@***.******@***.***&gt;;
主题: Re: [luow23/INP-Former] INP可视化 (Issue #28)


luow23 left a comment (luow23/INP-Former#28)
 
你好，我们用的解码器注意力图来可视化INP

**Comment by luow23** [comment]
Yes, you're right — we averaged the attention across the head dimension before performing the visualization.![Image](https://github.com/user-attachments/assets/9796b37d-e0bc-4dd5-a8a6-7b356d8c9b1b)

**Comment by shyyxy** [comment]
> 好的，感谢!
> […](#)

你好，请问你成功可视化了嘛
我用作者给的保存INPs的方法，从这个类`class Prototype_Block(nn.Module):`中返回注意力图到uad.py中编码器层，然后在 `def evaluation_batch`这个函数调用，为什么出来的图是这样

<img width="774" height="118" alt="Image" src="https://github.com/user-attachments/assets/a0698b5a-4753-470b-b24b-61ecb3690587" />

`   batch_num = imgs.shape[0]
   for i in range(batch_num):
       img_path_list = img_path[i].split('\\')
       class_name, category, idx_name = img_path_list[-4], img_path_list[-2], img_path_list[-1]
       # print(os.path.join(save_root, class_name, category))
       os.makedirs(os.path.join(save_root, class_name, category), exist_ok=True)
       input_frame = denormalize(imgs[i].clone().squeeze(0).cpu().detach().numpy())
       cv2_input = np.array(input_frame, dtype=np.uint8)
       plt.imsave(os.path.join(save_root, class_name, category, fr'{idx_name}_0.png'), cv2_input)
       ano_map = anomaly_map[i].squeeze(0).cpu().detach().numpy()
       plt.imsave(os.path.join(save_root, class_name, category, fr'{idx_name}_1.png'), ano_map, cmap='jet')
       gt_map = gt[i].squeeze(0).cpu().detach().numpy()
       plt.imsave(os.path.join(save_root, class_name, category, fr'{idx_name}_2.png'), gt_map, cmap='gray')
       distance = prototype_map[i].view((28, 28)).cpu().detach().numpy()
       distance = cv2.resize(distance, (392, 392), interpolation=cv2.INTER_AREA)
       plt.imsave(os.path.join(save_root, class_name, category, fr'{idx_name}_3.png'), distance, cmap='jet')
       attention = attention_map[i]
       prototype_num = attention.shape[0]
       for j in range(prototype_num):
           atten = attention[j]
           max_value = max(atten)
           min_value = min(atten)
           atten = (atten-min_value)/(max_value-min_value)
           atten = atten.view((28, 28)).cpu().detach().numpy()
           index = 3 + j + 1
           atten = cv2.resize(atten, (392, 392), interpolation=cv2.INTER_AREA)
           plt.imsave(os.path.join(save_root, class_name, category, fr'{idx_name}_{index}.png'), atten, cmap='jet')
       attention_sum = torch.sum(attention, dim=0).view((28, 28)).cpu().detach().numpy()
       plt.imsave(os.path.join(save_root, class_name, category, fr'{idx_name}_{4+prototype_num}.png'), attention_sum, cmap='hot')
       plt.close()`
如果您有空解答，我将不胜感激

**Comment by luow23** [comment]
attention = attention_list[4]，使用第五层的decoder layer进行可视化

**Comment by shyyxy** [comment]
> attention = attention_list[4]，使用第五层的decoder layer进行可视化

还是没有解决，你可以详细说一下嘛，非常感谢

**Comment by luow23** [comment]
decoder有8层，可视化第5层的attention试试看，或者其他层的

**Comment by shyyxy** [comment]
> decoder有8层，可视化第5层的attention试试看，或者其他层的可视化第五层吗
非常感谢，再次麻烦一下，我在uad.py的decoder加入下列代码，应该是可视化第五层的attention，顺便打印attention_map的min和max：
attention_map min: 0.0, max: 0.03501003235578537
attention_map shape: torch.Size([16, 6, 784])
attention_map min: 0.0, max: 0.03501003235578537
attention_map shape: torch.Size([16, 6, 784])
attention_map min: 0.0, max: 0.03501003235578537
attention_map shape: torch.Size([16, 6, 784])
attention_map min: 0.0, max: 0.03501003235578537
attention_map shape: torch.Size([16, 6, 784])
attention_map min: 0.0, max: 0.03501003235578537
`        de_list = []
        attention_list = []
        for i, blk in enumerate(self.decoder):
            # 调用时返回注意力图
            x, attn = blk(x, agg_prototype, return_attention=True)
            de_list.append(x)
            attention_list.append(attn)
        de_list = de_list[::-1]

        attention_list = [torch.mean(attn, dim=1).permute(0, 2, 1) for attn in attention_list]

        self.attention_map = attention_list[4]`
测试mvtec数据集第四层是可以可视化INPs的，但我用我的织物数据集每个层都试了，还是不行，只是保存这样的图

<img width="779" height="124" alt="Image" src="https://github.com/user-attachments/assets/d0fadadf-317c-4db1-a644-257d0163c3fe" />

**Comment by luow23** [comment]
你好，我觉得可能是织物数据集过于简单，纹理单一，只需要用一个INP就可以较好地建模，或者你可以试试INP-Former++中提出的soft inp coherence loss

**Comment by shyyxy** [comment]
> 好，我觉得可能是织物数据集过于简单，纹理单一，只需用一个INP就可以更好地建模，或者你可以试试INP-Former++中提出的软inp相干损失

非常感谢！
